src/lib/Libwin/passwd.c

```diff
@@ -3909,127 +3963,33 @@ LogonUserNoPass(char *user)
 	return (retval);
 }
 
-static HANDLE setuser_hdle = INVALID_HANDLE_VALUE;
-
-/* returns 0 for fail and non-zero for success */
+/* returns current user handle for fail and non-zero for success */
 /* records internally the current user handle */
-int
+HANDLE
 setuser(char *user)
 {
-	int	retval;
+	HANDLE setuser_hdle = INVALID_HANDLE_VALUE;
 
 	if (user == NULL)
-		return (0);
+		return setuser_hdle;
 
 	setuser_hdle = LogonUserNoPass(user);
-
-	if (setuser_hdle != INVALID_HANDLE_VALUE) {
-		retval = impersonate_user(setuser_hdle);
-	} else
-		retval =0;
-
-	return (retval);
-}
-
-
-/* setuser_with_password: like setuser() but with password cred_buf
- * and cred_len. If cred_buf is NULL, then it defaults to calling
- * setuser().
- * Returns: 0 for fail; non-zero for success
- */
-int
-setuser_with_password(char *user,
-	char *cred_buf,
-	size_t cred_len,
-	int (*decrypt_func)(char *, int, size_t, char **, const unsigned char *, const unsigned char *))
-{
-
-	SID     *usid;
-	char    thepass[PWLEN+1] = {'\0'};
-	char    realname[UNLEN+1] = {'\0'};
-	char    domain[PBS_MAXHOSTNAME+1] = "";
-	struct	passwd *pwdp = NULL;
-	int	i;
-
-	if ((user == NULL) || (decrypt_func == NULL))
-		return (0);
-
-	if (passwd_cache_init == 0) {
-		CLEAR_HEAD(passwd_cache_ll);
-		passwd_cache_init = 1;
-	}
-
-	/* look in internal cache for saved usertoken handle */
-	for (pwdp = (struct passwd*) GET_NEXT(passwd_cache_ll);
-		pwdp; pwdp = (struct passwd*) GET_NEXT(pwdp->pw_allpasswds)) {
-		if ( (strcmp(pwdp->pw_name, user) == 0) && \
-			(pwdp->pw_userlogin != INVALID_HANDLE_VALUE) ) {
-			setuser_hdle = pwdp->pw_userlogin;
-			return (impersonate_user(setuser_hdle));
-		}
-	}
-
-	usid = getusersid2(user, realname);
-	if (usid == NULL) {
-		return (0);
+	if (setuser_hdle == 0) {
+		log_err(-1, __func__, "failed in LogonUserNoPass");
```

Comment:
As per https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-logonusera
if the api fails then GetLastError gives the error code, which is what we are doing in log_err !!

Why do you think this is a bug ? if "setuser_hdle == 0" then it will ultimately return setuser_hdle

Reply:
But here the function LogonUserNoPass() used to set the "setuser_hdle" is not a windows api? more over looking at the implementation of LogonUserNoPass(), it is returning 0 at some place without setting last error. so I believe we should not call log_err(-1,), which might lead to log_err using a stale GetLastError() value which might be erroneous.

Sorry on the false BUG alarm! I failed to look down the lines 😄